[dtensor][cp][experiment] add CP experimental API to choose rotate method

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142093

**Summary**
This PR adds a new experimental API `set_rotate_method` for Context Parallel. This API allows user to choose the desired communication method (between all-to-all and all-gather) for shards rotation.

**Test**
`pytest test/distributed/_tensor/test_attention.py`

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

XilunWu committed:
[dtensor][cp][experiment] add CP experimental API to choose rotate method

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][cp][experiment] add CP experimental API to choose rotate method"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -1291,3 +1291,22 @@ def context_parallel_unshard(
+    Context Parallel SDPA requires the rotation of kv shards. Users can call this
+    API to specify which rotation method to use. "alltoall" shuffles the kv shards
+    using all-to-all collective while "allgather" gathers the kv shards using
+    all-gather collective. If this API has not been called, the default rotate
```
> while "allgather" gathers the kv shards using all-gather collective.

Can we change this to 

>. while "allgather" gathers the kv shards using all-gather collective after the first sub-SDPA computation.

We may want to implement the allgather version that happen before all sub-SDPA computation.

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -1291,3 +1291,22 @@ def context_parallel_unshard(
+    using all-to-all collective while "allgather" gathers the kv shards using
+    all-gather collective. If this API has not been called, the default rotate
+    method is "allgather".
+    """
```
Can we also add

```
Args:
    rotate_method(str): ...
```

I notice that I didn't do this for `context_parallel_unshard` either. We may want to fix it later. Sorry about the bad documentation.

fegin reviewed (approved):
LGTM. Some nits to fix before landing.

XilunWu committed:
Update on "[dtensor][cp][experiment] add CP experimental API to choose rotate method"


**Summary**
This PR adds a new experimental API `set_rotate_method` for Context Parallel. This API allows user to choose the desired communication method (between all-to-all and all-gather) for shards rotation.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][cp][experiment] add CP experimental API to choose rotate method"


**Summary**
This PR adds a new experimental API `set_rotate_method` for Context Parallel. This API allows user to choose the desired communication method (between all-to-all and all-gather) for shards rotation.

**Test**
`pytest test/distributed/_tensor/test_attention.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge

XilunWu commented:
@pytorchbot merge -f "unrelated CI failure (test_cuda)"
